docs: docs for OpenTelemetry instrumentation (#5291)
jina-ai/jina: docs/how-to/opentelemetry.md

```{hint}
The Jaeger provides a comprehensive out of the box tools for end-to-end tracing monitoring, visualization and alerting. Other tools can also be substituted to achieve the necessary goals of observability and performance analysis. The same can be said for Prometheus and Grafana.
```

### Docker Compose

A minimal docker-compose.yml file can look like:

```yaml
version: "3"
services:
  # Jaeger
  jaeger:
    image: jaegertracing/all-in-one:latest
    ports:
      - "16686:16686"

  otel-collector:
    image: otel/opentelemetry-collector:0.61.0
    command: [ "--config=/etc/otel-collector-config.yml" ]
    volumes:
      - ${PWD}/otel-collector-config.yml:/etc/otel-collector-config.yml
    ports:
      - "8888" # Prometheus metrics exposed by the collector
      - "8889" # Prometheus exporter metrics
      - "4317:4317" # OTLP gRPC receiver
    depends_on:
      - jaeger

  prometheus:
    container_name: prometheus
    image: prom/prometheus:latest
    volumes:
      - ${PWD}/prometheus-config.yml:/etc/prometheus/prometheus.yml
    ports:
      - "9090:9090"

  grafana:
    container_name: grafana
    image: grafana/grafana-oss:latest
    ports:
      - 3000:3000
```

> The corresponding OpenTelemetry Collector configuration below needs to be stored in file *otel-collector-config.yml*.

Reviewer: ```suggestion
The corresponding OpenTelemetry Collector configuration below needs to be stored in file `otel-collector-config.yml`:
```

```yaml
receivers:
  otlp:
    protocols:
      grpc:

exporters:
  jaeger:
    endpoint: jaeger:14250
    tls:
      insecure: true
  
  prometheus:
    endpoint: "0.0.0.0:8889"
    # can be used to add additional labels
    const_labels:
      label1: value1

processors:
  batch:

service:
  extensions: []
  pipelines:
    traces:
      receivers: [otlp]
      exporters: [jaeger]
      processors: [batch]
    metrics:
      receivers: [otlp]
      processors: [batch]
      exporters: [prometheus]
```

> Briefly, this setup creates a gRPC Collector Receiver on port 4317 that can collect data pushed by the Flow Pods. Collector exporters for Jaeger and Prometheus backends are configured to export tracing and metrics data respectively. The final **service** section creates a collector pipeline combining the receiver (collect data), export (to backend), process (batching) sub components.

Reviewer: ```suggestion
Briefly, this setup creates a gRPC Collector Receiver on port 4317 that can collect data pushed by the Flow Pods. Collector exporters for Jaeger and Prometheus backends are configured to export tracing and metrics data respectively. The final **service** section creates a collector pipeline combining the receiver (collect data) and exporter (to backend), process (batching) sub-components.
```

> The Prometheus minimal configuration below needs to be stored in file *prometheus-config.yml*.

Reviewer: ```suggestion
The Prometheus minimal configuration below needs to be stored in file `prometheus-config.yml`.
```
Backtick all filenames throughout PR please

```yaml
scrape_configs:
  - job_name: 'otel-collector'
    scrape_interval: 500ms
    static_configs:
      - targets: ['otel-collector:8889']
      - targets: ['otel-collector:8888']
```

> The Prometheus configuration now need only scrape from the OpenTelemetry Collector to get all the data from OpenTelemetry Metrics instrumented applications.

Reviewer: ```suggestion
The Prometheus configuration now only needs to  scrape from the OpenTelemetry Collector to get all the data from OpenTelemetry Metrics instrumented applications.
```

### Running a Flow locally

> Run the Flow and a sample request that we want to instrument locally. If the backends are running successfully the Flow has exported data to the Collector which can be queries and viewed which will be discussed next.

Reviewer: ```suggestion
Run the Flow and a sample request that we want to instrument locally. If the backends are running successfully the Flow has exported data to the Collector which can be queried and viewed.
```

```python
from jina import Flow, Document, DocumentArray

with Flow(
    tracing=True,
    traces_exporter_host='localhost',
    traces_exporter_port=4317,
    metrics=True,
    metrics_exporter_host='localhost',
    metrics_exporter_port=4317,
).add(uses='jinahub://SimpleIndexer') as f:
    f.post('/', DocumentArray([Document(text='hello')]))
```

## Viewing Traces in Jaeger UI

> The Jaeger UI can be opened by visiting this [url](http://localhost:16686). More details and guides on navigating the Jaeger UI can be found in the official [docs](https://www.jaegertracing.io/docs/1.38/external-guides/#using-jaeger).

Reviewer: ```suggestion
You can open the Jaeger UI [here](http://localhost:16686). You can find more information on the Jaeger UI in the official [docs](https://www.jaegertracing.io/docs/1.38/external-guides/#using-jaeger).
```

```{hint}
The list of available traces are documented in the {ref}`Flow Instrumentation <instrumenting-flow>` section.
```

## Monitor with Prometheus and Grafana

> External entities (like Grafana) can access these aggregated metrics via the query language [PromQL](https://prometheus.io/docs/prometheus/latest/querying/basics/) and let users visualize the metrics with dashboards. Comprehensive tutorial can be found [here](https://prometheus.io/docs/visualization/grafana/).

Reviewer: ```suggestion
External entities (like Grafana) can access these aggregated metrics via the query language [PromQL](https://prometheus.io/docs/prometheus/latest/querying/basics/), and let users visualize the metrics with dashboards. Check out a [comprehensive tutorial](https://prometheus.io/docs/visualization/grafana/) for more information.
```

```{hint}

> The list of available metrics are documented in the {ref}`Flow Instrumentation <instrumenting-flow>` section.

Reviewer: ```suggestion
A list of available metrics is in the {ref}`Flow Instrumentation <instrumenting-flow>` section.
```

```

## JCloud Support

> Currently OpenTelemetry is not supported by JCloud. The features will be made available soon. Until then the deprecated Prometheus based {ref}`monitoring setup <monitoring-flow>` can be used.

Reviewer: ```suggestion
JCloud doesn't currently support OpenTelemetry. These features will be made available soon. Until then, you can use the deprecated Prometheus-based {ref}`monitoring setup <monitoring-flow>`.
```

(opentelemetry)=
# Observability and Instrumentation with OpenTelemetry and Jina

```{hint}

> Prometheus only based metrics collection will be deprecated soon. Refer to {ref}`Monitor with Prometheus and Grafana <monitoring>` for the old setup.

Reviewer 2: I might be wrong, but I think it should be "Prometheus-only" with the `-`. @alexcg1 can you help me on this one?

Reviewer: Yup, hyphen needed

```

> There are two major setups required to visualize/monitor your application's signals using [OpenTelemetry](https://opentelemetry.io). The first step is covered by Jina which integrates the [OpenTelemetry API and SDK](https://opentelemetry-python.readthedocs.io/en/stable/api/index.html) at the application level. The {ref}`Flow Instrumentation <instrumenting-flow>` page covers in detail the steps required to enable OpenTelemetry in a Flow. A {class}`~jina.Client` can also be instrumented which is documented in the {ref}`Client Instrumentation <instrumenting-client>` section.

Reviewer: - "There are two major setups"
- "The first step"

Step or Setup?

This section covers the OpenTelemetry infrastructure setup required to collect, store and visualize the traces and metrics data exported by the Pods. This setup is the user's responsibility, and this section only serves as the initial/introductory guide to running OpenTelemetry infrastructure components.

> Since OpenTelemetry is open source and is mostly responsible for the API standards and specification, there are various providers that implement the specification. This section follows the default recommendations from the OpenTelemetry documentation that also fits into the Jina implementations.

Reviewer: ```suggestion
Since OpenTelemetry is open source and is mostly responsible for the API standards and specification, various providers implement the specification. This section follows the default recommendations from the OpenTelemetry documentation that also fits into the Jina implementations.
```

## Exporting traces and metrics data

Pods created using a {class}`~jina.Flow` with tracing or metrics enabled use the [SDK Exporters](https://opentelemetry.io/docs/instrumentation/python/exporters/) to send the data to a central [Collector](https://opentelemetry.io/docs/collector/) component. You can use this collector to further process and store the data for visualization and alerting. 

The push/export-based mechanism also allows the application to start pushing data immediately on startup. This differs from the pull-based mechanism where you need a separate scraping registry to discovery service to identify data scraping targets.

> You can currently configure the exporter backend host and port using the `traces_exporter_host`, `traces_exporter_port`, `metrics_exporter_host` and `metrics_exporter_port`. Even though the Collector is metric data type agnostic (it can accept any type of OpenTelemetry API data model), we provide separate configuration for Tracing and Metrics to give users more flexibility in choosing their infrastructure components.

Reviewer: ```suggestion
You can configure the exporter backend host and port using the `traces_exporter_host`, `traces_exporter_port`, `metrics_exporter_host` and `metrics_exporter_port`. Even though the Collector is metric data-type agnostic (it accepts any type of OpenTelemetry API data model), we provide separate configuration for Tracing and Metrics to give you more flexibility in choosing infrastructure components.
```

> Jina's default exporter implementation is  `OTLPSpanExporter` and `OTLPMetricExporter`. The exporters also use the gRPC data transfer protocol. The following environment variables can be used to further configure the exporter client based on your requirements. The full list of exporter related environment variables are documented by the [PythonSDK library](https://opentelemetry-python.readthedocs.io/en/latest/exporter/otlp/otlp.html). Apart from the `OTEL_EXPORTER_OTLP_PROTOCOL` and the `OTEL_EXPORTER_OTLP_TRACES_ENDPOINT` environment variables, you can use all other library version specific environment variables to configure the exporter clients.

Reviewer: ```suggestion
Jina's default exporter implementation is  `OTLPSpanExporter` and `OTLPMetricExporter`. The exporters also use the gRPC data transfer protocol. The following environment variables can be used to further configure the exporter client based on your requirements. The full list of exporter related environment variables are documented by the [PythonSDK library](https://opentelemetry-python.readthedocs.io/en/latest/exporter/otlp/otlp.html). Apart from `OTEL_EXPORTER_OTLP_PROTOCOL` and `OTEL_EXPORTER_OTLP_TRACES_ENDPOINT`, you can use all other library version specific environment variables to configure the exporter clients.
```

## Collector

> The [Collector](https://opentelemetry.io/docs/collector/) is a huge ecosystem of components that support various features such as scraping, collecting, processing and further exporting data to storage backends. The collector itself can also expose endpoints to allow scraping data. We recommend you read the official documentation to understand the the full set of features and configuration required to run a Collector. Read the below section to understand the minimum number of components and the respective configuration required for operating with Jina.

Reviewer: ```suggestion
The [Collector](https://opentelemetry.io/docs/collector/) is a huge ecosystem of components that support features like scraping, collecting, processing and further exporting data to storage backends. The collector itself can also expose endpoints to allow scraping data. We recommend reading the official documentation to understand the the full set of features and configuration required to run a Collector. Read the below section to understand the minimum number of components and the respective configuration required for operating with Jina.
```

      - ${PWD}/otel-collector-config.yml:/etc/otel-collector-config.yml
    ports:
      - "8888" # Prometheus metrics exposed by the collector
      - "8889" # Prometheus exporter metrics
      - "4317:4317" # OTLP gRPC receiver
    depends_on:
      - jaeger

  prometheus:
    container_name: prometheus
    image: prom/prometheus:latest
    volumes:
      - ${PWD}/prometheus-config.yml:/etc/prometheus/prometheus.yml
    ports:
      - "9090:9090"

  grafana:
    container_name: grafana
    image: grafana/grafana-oss:latest
    ports:
      - 3000:3000
```

The corresponding OpenTelemetry Collector configuration below needs to be stored in file `otel-collector-config.yml`:
```yaml
receivers:
  otlp:
    protocols:
      grpc:

exporters:
  jaeger:
    endpoint: jaeger:14250
    tls:
      insecure: true
  
  prometheus:
    endpoint: "0.0.0.0:8889"
    # can be used to add additional labels
    const_labels:
      label1: value1

processors:
  batch:

service:
  extensions: []
  pipelines:
    traces:
      receivers: [otlp]
      exporters: [jaeger]
      processors: [batch]
    metrics:
      receivers: [otlp]
      processors: [batch]
      exporters: [prometheus]
```

> Briefly, this setup creates a gRPC Collector Receiver on port 4317 that can collect data pushed by the Flow Pods. Collector exporters for Jaeger and Prometheus backends are configured to export tracing and metrics data respectively. The final **service** section creates a collector pipeline combining the receiver (collect data) and exporter (to backend), process (batching) sub-components.

Reviewer: ```suggestion
This setup creates a gRPC Collector Receiver on port 4317 that collects data pushed by the Flow Pods. Collector exporters for Jaeger and Prometheus backends are configured to export tracing and metrics data respectively. The final **service** section creates a collector pipeline combining the receiver (collect data) and exporter (to backend), process (batching) sub-components.
```

> The Prometheus minimal configuration below needs to be stored in file `prometheus-config.yml`.

Reviewer: ```suggestion
The minimal Prometheus configuration needs to be stored in `prometheus-config.yml`.
```

      - targets: ['otel-collector:8888']
```

The Prometheus configuration now only needs to scrape from the OpenTelemetry Collector to get all the data from OpenTelemetry Metrics instrumented applications.

### Running a Flow locally

Run the Flow and a sample request that we want to instrument locally. If the backends are running successfully the Flow has exported data to the Collector which can be queried and viewed.

```python
from jina import Flow, Document, DocumentArray

with Flow(
    tracing=True,
    traces_exporter_host='localhost',
    traces_exporter_port=4317,
    metrics=True,
    metrics_exporter_host='localhost',
    metrics_exporter_port=4317,
).add(uses='jinahub://SimpleIndexer') as f:
    f.post('/', DocumentArray([Document(text='hello')]))
```

## Viewing Traces in Jaeger UI

You can open the Jaeger UI [here](http://localhost:16686). You can find more information on the Jaeger UI in the official [docs](https://www.jaegertracing.io/docs/1.38/external-guides/#using-jaeger).

```{hint}
The list of available traces are documented in the {ref}`Flow Instrumentation <instrumenting-flow>` section.
```

## Monitor with Prometheus and Grafana

> External entities (like Grafana) can access these aggregated metrics via the query language [PromQL](https://prometheus.io/docs/prometheus/latest/querying/basics/), and let users visualize the metrics with dashboards. Check out a [comprehensive tutorial](https://prometheus.io/docs/visualization/grafana/) for more information.

Reviewer: ```suggestion
External entities (like Grafana) can access these aggregated metrics via the [PromQL](https://prometheus.io/docs/prometheus/latest/querying/basics/) query language, and let users visualize metrics with dashboards. Check out a [comprehensive tutorial](https://prometheus.io/docs/visualization/grafana/) for more information.
```

> You can download a [sample Grafana dashboard JSON file](https://github.com/jina-ai/example-grafana-prometheus/blob/main/grafana-dashboards/flow-histogram-metrics.json) and import it into Grafana to get started with some pre-built graphs.
> 
> The sample dashboard looks like:

Reviewer: ```suggestion
Download a [sample Grafana dashboard JSON file](https://github.com/jina-ai/example-grafana-prometheus/blob/main/grafana-dashboards/flow-histogram-metrics.json) and import it into Grafana to get started with some pre-built graphs:
```

```{figure} ../../.github/2.0/grafana-histogram-metrics.png
:align: center
```

```{hint}
:class: seealso
A list of available metrics is in the {ref}`Flow Instrumentation <instrumenting-flow>` section.
To update your existing Prometheus and Grafana configurations, refer to the {ref}`OpenTelemetry migration guide <opentelemetry-migration>`.
```

## JCloud Support

> JCloud doesn't currently support OpenTelemetry. These features will be made available soon. Until then, you can use the deprecated Prometheus-based {ref}`monitoring setup <monitoring-flow>`.

Reviewer: ```suggestion
JCloud doesn't currently support OpenTelemetry. We'll make these features available soon. Until then, you can use the deprecated Prometheus-based {ref}`monitoring setup <monitoring-flow>`.
```

External entities (like Grafana) can access these aggregated metrics via the query language [PromQL](https://prometheus.io/docs/prometheus/latest/querying/basics/), and let users visualize the metrics with dashboards. Check out a [comprehensive tutorial](https://prometheus.io/docs/visualization/grafana/) for more information.

> A sample Grafana dashboard JSON file can downloaded from [here](https://github.com/jina-ai/example-grafana-prometheus/blob/main/grafana-dashboards/flow-histogram-metrics.json). The JSON file can be imported into Grafana to get started with some pre-built graphs.

Reviewer: ```suggestion
You can download a [sample Grafana dashboard JSON file](https://github.com/jina-ai/example-grafana-prometheus/blob/main/grafana-dashboards/flow-histogram-metrics.json) and import it into Grafana to get started with some pre-built graphs.
```

> A sample Grafana dashboard JSON file can downloaded from [here](https://github.com/jina-ai/example-grafana-prometheus/blob/main/grafana-dashboards/flow-histogram-metrics.json). The JSON file can be imported into Grafana to get started with some pre-built graphs.

Reviewer: See prior comment on JSON file and adapt this part accordingly

```{hint}
:class: seealso
A list of available metrics is in the {ref}`Flow Instrumentation <instrumenting-flow>` section.

> Refer to the {ref}`OpenTelemetry migration guide <opentelemetry-migration>` for updating your existing Prometheus and Grafana configurations.

Reviewer: ```suggestion
To update your existing Prometheus and Grafana configurations, refer to the {ref}`OpenTelemetry migration guide <opentelemetry-migration>`.
```
